File: core/player/ships/player.js
import { audioCtx } from "../../assets/audio/context.js";
import { assets } from "../../assets/main.js";
import { clamp } from "../../utils/math.js";
import { randomNum } from "../../utils/random.js";
import Ship from "./ship.js";


export let ship;

export default class PlayerShip extends Ship {
    constructor({ x, y, width, height, angle, acceleration, color = "red", name = "Player", controllable = true, turnRate, maxWeaponHeat = 10000 }) {
        super({ x, y, width, height, angle, img: assets?.images?.mainship, flameImg: assets?.images?.flame1, acceleration, color, name, maxWeaponHeat, controllable, life: 10000, turnRate });
        this.audioGain = audioCtx.createGain();

        // this.audioPlayer = throtlePlayAudio(assets?.audios?.engine, this.audioGain);

        if (this.controllable === false)
            this.state = "idle";
    }

    update(t, dt) {
        super.update(t, dt);


        if (this.speed > 50) {
            const { maxValue, minValue } = this.audioGain.gain;
            this.audioGain.gain.value = clamp(this.speed / 200, minValue, maxValue);
            // this.audioPlayer();
        }
    }

    static spawn(x, y) {
        const angle = randomNum(-Math.PI * 2, Math.PI * 2);
        const spawnDistance = 200000;
        const spawnX = randomNum(x - spawnDistance, x + spawnDistance);
        const spawnY = randomNum(y - spawnDistance, y + spawnDistance);
        ship = new PlayerShip({ x: spawnX, y: spawnY, angle, width: 50, height: 50, color: "#84d0ff", controllable: true, acceleration: 600, turnRate: 4.5, maxWeaponHeat: 10000 });
    }
}    


// 600-1500
// 2-4.5

File: void_front/game/core/player/ships/player.py
from __future__ import annotations

from game.core.assets.audio.context import audioCtx
from game.core.assets.main import assets
from game.core.utils.math import clamp
from game.core.player.ships.ship import Ship
import numpy as np

ship = None


class PlayerShip(Ship):
    def __init__(self, *, x, y, width, height, angle, acceleration=1000, color="red", name="Player", controllable=True, turnRate=2, maxWeaponHeat=10000):
        super().__init__(x=x, y=y, width=width, height=height, angle=angle, img=getattr(getattr(assets, "images", None), "mainship", None), flameImg=getattr(getattr(assets, "images", None), "flame1", None), acceleration=acceleration, color=color, name=name, maxWeaponHeat=maxWeaponHeat, controllable=controllable, life=10000, turnRate=turnRate)
        self.audioGain = audioCtx.createGain()
        if self.controllable is False:
            self.state = "idle"

    def update(self, t, dt, thrust=0, turn=0):
        super().update(t, dt, thrust=thrust, turn=turn)
        if self.speed > 50:
            gain = self.audioGain.gain
            self.audioGain.gain.value = clamp(self.speed / 200, gain.minValue, gain.maxValue)

    @staticmethod
    def spawn(x, y):
        global ship
        angle = np.random.uniform(-np.pi * 2, np.pi * 2)
        spawnDistance = 200000
        spawnX = np.random.uniform(x - spawnDistance, x + spawnDistance)
        spawnY = np.random.uniform(y - spawnDistance, y + spawnDistance)
        ship = PlayerShip(x=spawnX, y=spawnY, angle=angle, width=50, height=50, color="#84d0ff", controllable=False, acceleration=np.random.uniform(600, 1500), turnRate=np.random.uniform(2, 4.5))
        
        return ship

